### PR TITLE
Internal: Added Onboarding Playwright tests [ED-8799]

### PR DESCRIPTION
**Pull request by nickk-elementor**
## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [X] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry: 

* Internal: Added Onboarding Playwright tests

## Quality assurance

- [X] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: tests/playwright/sanity/core/app/modules/onboarding/onboarding.data.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+    "createAccountButtonText" : "Create my account",
+
```
Comment (udidol) [inline tests/playwright/sanity/core/app/modules/onboarding/onboarding.data.json:3]: Remove all empty lines, add empty line at the end after the closing curly brace

File: tests/playwright/sanity/core/app/modules/onboarding/onboarding.test.js [hunk 2]
```diff
@@ -1,144 +1,181 @@
...
+
+		const [ importPopup ] = await Promise.all( [
+			page.waitForEvent( 'popup' ),
+			helper.selectAlreadyHaveElementorProLink(),
+		] );
+
+		await expect( await importPopup.url() ).toContain( 'uploadAndInstallPro' );
+	} );
+
+	test( '"Connect Your Account" link works', async ( { page } ) => {
```
Comment (udidol) [inline tests/playwright/sanity/core/app/modules/onboarding/onboarding.test.js:39]: If you don't use `page`, see if you can just remove it from the test (in all places)

File: tests/playwright/config/global-setup.js [hunk 3]
```diff
@@ -6,11 +6,11 @@ module.exports = async ( config ) => {
 	const browser = await chromium.launch( { headless: config.headless } ),
 		page = await browser.newPage();
 
-	await page.goto( `${ config.baseURL }/wp-admin` );
+	await page.goto( `${ config.baseURL }/wp-admin`, { waitUntil: 'networkidle' } );
```
Comment (hein-obox) [inline tests/playwright/config/global-setup.js:9]: Is it possible to undo your changes in this file? Because this will affect all users.

File: tests/playwright/pages/wp-admin-page.js [hunk 4]
```diff
@@ -23,8 +23,8 @@ module.exports = class WpAdminPage extends BasePage {
 		}
 
 		await this.page.waitForSelector( 'text=Log In' );
-		await this.page.fill( 'input[name="log"]', this.config.user.username );
-		await this.page.fill( 'input[name="pwd"]', this.config.user.password );
+		await this.page.type( 'input[name="log"]', this.config.user.username, { delay: 100 } );
+		await this.page.type( 'input[name="pwd"]', this.config.user.password, { delay: 100 } );
```
Comment (hein-obox) [inline tests/playwright/pages/wp-admin-page.js:27]: Is it possible to undo this change? Because this will affect all users.

File: tests/playwright/pages/wp-admin-page.js [hunk 5]
```diff
@@ -105,4 +105,28 @@ module.exports = class WpAdminPage extends BasePage {
 		await this.page.selectOption( '#WPLANG', language );
 		await this.page.locator( '#submit' ).click();
 	}
+
+	async gotoThemesPage() {
+        await this.page.goto( '/wp-admin/themes.php', { waitUntil: 'networkidle' } );
```
Comment (hein-obox) [inline tests/playwright/pages/wp-admin-page.js:110]: Suggestion: can you create a function called `activateHelloElementorTheme()` or `activateTheme( 'Hello Elementor' )`

File: tests/playwright/pages/wp-admin-page.js [hunk 6]
```diff
@@ -105,4 +105,28 @@ module.exports = class WpAdminPage extends BasePage {
...
+    }
+
+	async activateHelloTheme() {
+        await this.gotoThemesPage();
+		if ( await this.page.locator( '[aria-label="Activate Hello Elementor"]' ).isVisible() ) {
```
Comment (hein-obox) [inline tests/playwright/pages/wp-admin-page.js:115]: Is it possible to use a classname or ID for this, to make it language agnostic?

File: tests/playwright/sanity/core/app/modules/onboarding/onboarding.test.js [hunk 7]
```diff
@@ -1,146 +1,176 @@
...
-
-	await expect( goProPopover ).toBeVisible();
+const { onboarding } = require( './onboarding.utils' );
+const WpAdminPage = require( '../../../../../pages/wp-admin-page' );
+const testData = JSON.parse( JSON.stringify( require( './onboarding.data.json' ) ) );
+
+test.beforeAll( async ( { browser }, testInfo ) => {
+	const page = await browser.newPage();
+	helper = new onboarding( page, testInfo.project.use.baseURL );
+	wpAdminHelper = new WpAdminPage( page, testInfo );
```
Comment (hein-obox) [inline tests/playwright/sanity/core/app/modules/onboarding/onboarding.test.js:9]: In other tests we call this `wpAdminPage` I believe.

File: tests/playwright/pages/wp-admin-page.js [hunk 8]
```diff
@@ -105,4 +105,28 @@ module.exports = class WpAdminPage extends BasePage {
...
+            await this.page.locator( '[data-slug="hello-elementor"]' ).hover();
+
+            await this.page.locator( '[aria-label="Activate Hello Elementor"]' ).click();
+            await this.page.waitForLoadState( 'networkidle' );
+        }
+    }
+
+	async activateTwenty21Theme() {
```
Comment (hein-obox) [inline tests/playwright/pages/wp-admin-page.js:123]: Please indentation.

File: tests/playwright/sanity/core/app/modules/onboarding/onboarding.data.json [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+    "createAccountButtonText" : "Create my account",
+    "disclaimerNotice" : "* You can switch your theme later on",
+    "importElementorPopupTitle" : "Elementor | Upload and Install Elementor Pro",
+    "signIntoElementorPopupTitle" : "Sign in – My Account",
+    "upgradeFeaturesList" : [
+        "90+ Basic & Pro widgets",
+        "300+ Basic & Pro templates",
+        "Premium Support"
+    ]
+}
```
Comment (hein-obox) [inline tests/playwright/sanity/core/app/modules/onboarding/onboarding.data.json:11]: Please add new line at the end of the file.

File: tests/playwright/sanity/core/app/modules/onboarding/onboarding.utils.js [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+
+        // Validate Broken CSS Requests
+        await expect( requestFailed, `Failed Assets: ${ requestFailed.toString() }` ).toEqual( [] );
+    }
+
+    async gotoExperiments() {
+        await this.page.goto( '/wp-admin/admin.php?page=elementor#tab-experiments', { waitUntil: 'networkidle' } );
+    }
+
+    async checkflexBoxIsOff() {
```
Comment (hein-obox) [inline tests/playwright/sanity/core/app/modules/onboarding/onboarding.utils.js:85]: Can we use the existing `setExperiments()` function for that?

File: tests/playwright/sanity/core/app/modules/onboarding/onboarding.utils.js [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+			// Opens popup.
+			this.selectCreateMyAccountCTA(),
+		] );
+        await expect( await createMyAccountPopUp.title(), `"Sign up for Elementor" text is not present on the pop up` ).toEqual( 'Sign Up – My Account' );
+        await createMyAccountPopUp.close();
+    }
+
+    async alreadyHaveElementorProWorks() {
+        const [ alreadyHaveElementorProPopup ] = await Promise.all( [
+			this.page.waitForEvent( 'popup' ),
```
Comment (hein-obox) [inline tests/playwright/sanity/core/app/modules/onboarding/onboarding.utils.js:208]: Indentation?

File: tests/playwright/sanity/core/app/modules/onboarding/onboarding.utils.js [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+    }
+
+    async checkUserIsOnStepTwo() {
+        await expect( this.page.url(), `User is not on Step 2 but instead on ${ await this.page.url() }` ).toEqual( this.baseUrl + this.step2URL );
+    }
+
+    async selectContinueWithHelloThemeButton() {
+        await this.continueWithHelloThemeButton.click();
+        await this.page.waitForLoadState( 'networkidle' );
+        await this.page.waitForTimeout( 250 );
```
Comment (hein-obox) [inline tests/playwright/sanity/core/app/modules/onboarding/onboarding.utils.js:250]: Is there a way to use a waitForElement kind of selector instead of a timeout?

File: tests/playwright/sanity/core/app/modules/onboarding/onboarding.utils.js [hunk 13]
```diff
@@ -0,0 +1,438 @@
...
+        await this.page.waitForLoadState( 'networkidle' );
+        await this.page.waitForSelector( '#customize-control-custom_logo .upload-button' );
+    }
+
+    async uploadLogo() {
+        await this.page.locator( '#customize-control-custom_logo .upload-button' ).first().click();
+        await this.page.waitForLoadState( 'networkidle' );
+        await this.page.locator( '#menu-item-browse' ).click();
+        await this.page.waitForLoadState( 'networkidle' );
+        await this.page.waitForTimeout( 500 );
```
Comment (hein-obox) [inline tests/playwright/sanity/core/app/modules/onboarding/onboarding.utils.js:317]: Can you compare the upload process of images e.g. for the Image Carousel or Image Gallery, to see it we can find a way to avoid the timeout?

File: tests/playwright/sanity/core/app/modules/onboarding/onboarding.utils.js [hunk 14]
```diff
@@ -0,0 +1,438 @@
...
+          ] );
+          await fileChooser.setFiles( 'logo.png' );
+        }
+    }
+
+    async addALogo() {
+        await this.openMediaLibraryButton.click();
+        await this.page.waitForLoadState( 'networkidle' );
+        await this.page.locator( '#menu-item-browse' ).click();
+        await this.page.waitForTimeout( 500 );
```
Comment (hein-obox) [inline tests/playwright/sanity/core/app/modules/onboarding/onboarding.utils.js:389]: Timeout?

File: tests/playwright/sanity/core/app/modules/onboarding/onboarding.utils.js [hunk 15]
```diff
@@ -0,0 +1,438 @@
...
+            this.kitLibraryButton.click(),
+          ] );
+    }
+
+    async checkUserIsOnTheLibraryKitsPage() {
+        await expect( await this.page.url().includes( '/kit-library' ), `The User is on: ${ this.page.url() }` ).toBeTruthy();
+    }
+
+    async checkFirstKitIsBlankCanvas() {
+        await this.page.waitForTimeout( 1000 );
```
Comment (hein-obox) [inline tests/playwright/sanity/core/app/modules/onboarding/onboarding.utils.js:420]: Do we need the timeout?

File: tests/playwright/sanity/image-carousel.test.js [hunk 16]
```diff
@@ -3,8 +3,9 @@ const WpAdminPage = require( '../pages/wp-admin-page.js' );
 
 test( 'Image Carousel', async ( { page }, testInfo ) => {
 	// Arrange.
-	const wpAdmin = new WpAdminPage( page, testInfo ),
-		editor = await wpAdmin.useElementorCleanPost();
+	const wpAdmin = new WpAdminPage( page, testInfo );
```
Comment (hein-obox) [inline tests/playwright/sanity/image-carousel.test.js:6]: Please restore the original version. Try not to make change if not necessary.

File: tests/playwright/sanity/image-carousel.test.js [hunk 17]
```diff
@@ -21,7 +22,11 @@ test( 'Image Carousel', async ( { page }, testInfo ) => {
   // Populate the widget with images.
   await editor.populateImageCarousel();
 
-  expect( await editor.getPreviewFrame().locator( 'div.elementor-image-carousel-wrapper.swiper-container.swiper-container-initialized' ).screenshot( { type: 'jpeg', quality: 70 } ) ).toMatchSnapshot( 'carousel.jpeg' );
+  await editor.togglePreviewMode();
```
Comment (hein-obox) [inline tests/playwright/sanity/image-carousel.test.js:25]: For reviewer 2: 
Nik and I have made this change to make this test less flaky.

**Review by hein-obox — COMMENTED** [review]
See my feedback.

File: tests/playwright/config/global-setup.js (hunk 3 above)
Comment (nickk-elementor) [reply tests/playwright/config/global-setup.js:9]: We discussed with @Nevoss that this would be a good solution. 
Often when running the tests locally it fails to log into Wordpress. So I change it every time locally to consistently have the tests run.

File: tests/playwright/pages/wp-admin-page.js (hunk 6 above)
Comment (nickk-elementor) [reply tests/playwright/pages/wp-admin-page.js:115]: This button only contains an aria-label with the theme name. I could not find a way to create a selector where you just change out the ID or class name. So now I made it one function and you can pass in any selector to activate it.

**Comment by hein-obox** [comment]
I am thinking could you ask @Nevo Golan ***@***.***> to approve
this change or to do PR2?

File: tests/playwright/config/global-setup.js (hunk 3 above)
Comment (nickk-elementor) [reply tests/playwright/config/global-setup.js:9]: I tried to replace 'networkidle' with 'load' and the tests pass the authentication on WP every other time.

File: tests/playwright/sanity/core/app/modules/onboarding/onboarding.data.json (hunk 9 above)
Comment (hein-obox) [reply tests/playwright/sanity/core/app/modules/onboarding/onboarding.data.json:11]: The extra line hasn't been added yet. 

Can you please not resolve conversations? I prefer to resolve the conversations which I started as a reviewer.

File: tests/playwright/sanity/image-carousel.test.js (hunk 16 above)
Comment (hein-obox) [inline tests/playwright/sanity/image-carousel.test.js:6]: The indentation seems a bit odd.

File: tests/playwright/sanity/image-carousel.test.js [hunk 18]
```diff
@@ -3,8 +3,9 @@ const WpAdminPage = require( '../pages/wp-admin-page.js' );
 
 test( 'Image Carousel', async ( { page }, testInfo ) => {
 	// Arrange.
-	const wpAdmin = new WpAdminPage( page, testInfo ),
-		editor = await wpAdmin.useElementorCleanPost();
+	const wpAdmin = new WpAdminPage( page, testInfo );
+
+  editor = await wpAdmin.useElementorCleanPost();
```
Comment (hein-obox) [inline tests/playwright/sanity/image-carousel.test.js:8]: shouldn't the `editor` have `const editor` and can't we remove the empty line? Compare other tests.

File: tests/playwright/pages/wp-admin-page.js (hunk 4 above)
Comment (hein-obox) [reply tests/playwright/pages/wp-admin-page.js:27]: Do we need the `delay`? I have never used it, and it worked fine without.
